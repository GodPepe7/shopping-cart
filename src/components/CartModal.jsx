import Close from "../assets/icons/close.svg";
import { useRef, useEffect } from "react";
import { useOutletContext } from "react-router-dom";
import BasketItem from "./BasketItem";

function CartModal({ modalIsOpen, closeModalHandler }) {
  const { basket } = useOutletContext();
  const ref = useRef(null);
  useEffect(() => {
    if (modalIsOpen) {
      ref.current?.showModal();
    } else {
      ref.current?.close();
    }
  }, [modalIsOpen]);

  return (
    <dialog
      className="min-h-screen min-w-[250px] w-[20%] backdrop:backdrop-blur px-4 py-4 ml-auto mr-0"
      ref={ref}
      onCancel={closeModalHandler}
    >
      <button onClick={closeModalHandler} className="w-5 h-5">
        <img src={Close} alt="Close Shopping Cart" />
      </button>
      <h2 className="text-center mb-4">Your Basket</h2>
      <div>
          {[...basket.entries()].map(
            ([productId, { title, coverImage, price, quantity }]) => (
              <BasketItem
                key={productId}
                title={title}
                coverImage={coverImage}
                price={price}
                quantity={quantity}
              ></BasketItem>
            )
          )}
          <div className=" text-xl my-2">
            <p>Total Price</p>
            <p>
              {[...basket.values()].reduce(
                (accumulator, { price, quantity }) =>
                  accumulator + quantity * price,
                0
              )}
              â‚¬
            </p>
          </div>
          <button className="py-1 w-full text-theme-white bg-theme-black shadow-md uppercase tracking-wider">
            Checkout
          </button>
      </div>
    </dialog>
  );
}

export default CartModal;
