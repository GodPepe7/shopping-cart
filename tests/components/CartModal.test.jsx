/* eslint-disable no-undef */
import { vi } from "vitest";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import CartModal from "../../src/components/CartModal";
import * as rrd from "react-router-dom";
import { testData } from "../testData";

beforeEach(() => {
  // Vitest which uses jsdom currently doesn't support
  // the HTMLDialogElement with its functions so for now I had to mock them
  // https://github.com/jsdom/jsdom/issues/3294
  HTMLDialogElement.prototype.show = vi.fn();
  HTMLDialogElement.prototype.showModal = vi.fn();
  HTMLDialogElement.prototype.close = vi.fn();
});

describe("CartModal", () => {
  vi.mock("react-router-dom");
  const mockSetCart = vi.fn();
  rrd.useOutletContext.mockReturnValue({
    basket: new Map(),
    setBasket: () => mockSetCart,
  });

  it("Matches Snapshot", () => {
    const opened = false;
    const onClick = vi.fn();

    const container = render(
      <CartModal modalIsOpen={opened} closeModalHandler={onClick} />
    );

    expect(container).toMatchSnapshot();
  });

  it("Displays cart is empty when no items in shopping cart", () => {
    const opened = false;
    const onClick = vi.fn();

    render(<CartModal modalIsOpen={opened} closeModalHandler={onClick} />);
    expect(screen.getByText("Shopping Cart is empty")).toBeInTheDocument();
  });

  it("Call onClick when pressing close button", async () => {
    const user = userEvent.setup();
    const opened = true;
    const onClick = vi.fn();

    render(<CartModal modalIsOpen={opened} closeModalHandler={onClick} />);
    // cause the modal isn't being shown due to jsdom not supporting
    // HTMLDialogElement yet, resulting in the modal staying hidden
    // hidden: true has to be set so getByRole can find the button
    const closeBtn = screen.getByRole("button", {
      name: /close/i,
      hidden: true,
    });
    await user.click(closeBtn);

    expect(onClick).toHaveBeenCalled();
  });

  it("Render 2 basket items with correct total price and checkout button", async () => {
    const opened = true;
    const onClick = vi.fn();
    rrd.useOutletContext.mockReturnValue({
      basket: testData,
      setBasket: () => vi.fn(),
    });

    render(<CartModal modalIsOpen={opened} closeModalHandler={onClick} />);
    const cartItemOne = screen.getByRole("heading", {
      name: /chainsaw man/i,
      hidden: true,
    });
    const cartItemTwo = screen.getByRole("heading", {
      name: /attack on titan/i,
      hidden: true,
    });
    const totalPrice = screen.getByText("14â‚¬");
    const checkoutBtn = screen.getByRole("button", {
      name: /checkout/i,
      hidden: true,
    });

    expect(totalPrice).toBeInTheDocument();
    expect(checkoutBtn).toBeInTheDocument();
    expect(cartItemOne).toBeInTheDocument();
    expect(cartItemTwo).toBeInTheDocument();

    screen.debug();
  });
});
