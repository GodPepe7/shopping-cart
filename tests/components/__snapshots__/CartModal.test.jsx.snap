// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CartModal > Matches Snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <dialog
        class="min-h-screen min-w-[250px] w-[40%] max-w-[400px] backdrop:backdrop-blur px-4 py-4 ml-auto mr-0 animate-[slide-out_0.15s_linear]"
      >
        <button
          aria-label="close"
          class="w-5 h-5"
        >
          <img
            alt="Close Shopping Cart"
            src="/src/assets/icons/close.svg"
          />
        </button>
        <h2
          class="text-center mb-4"
        >
          Your Basket
        </h2>
        <p
          class="text-center italic font-extralight"
        >
          Shopping Cart is empty
        </p>
      </dialog>
    </div>
  </body>,
  "container": <div>
    <dialog
      class="min-h-screen min-w-[250px] w-[40%] max-w-[400px] backdrop:backdrop-blur px-4 py-4 ml-auto mr-0 animate-[slide-out_0.15s_linear]"
    >
      <button
        aria-label="close"
        class="w-5 h-5"
      >
        <img
          alt="Close Shopping Cart"
          src="/src/assets/icons/close.svg"
        />
      </button>
      <h2
        class="text-center mb-4"
      >
        Your Basket
      </h2>
      <p
        class="text-center italic font-extralight"
      >
        Shopping Cart is empty
      </p>
    </dialog>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
